File: ios/Sources/UnifiedTrackingPlugin/ErrorTracking/RollbarProvider.swift
import Foundation
import RollbarNotifier

/// Rollbar (rollbar-ios / RollbarNotifier) bridge for the unified-tracking plugin.
///
/// Reads `accessToken` (required) and optional `environment` from the provider
/// config and initializes the Rollbar SDK. If no access token is supplied the
/// SDK is not started and the provider stays uninitialized.
///
/// NOTE: Rollbar does NOT expose a first-class consent toggle. When consent is
/// revoked we gate `logError` at the provider level (`consentGranted`). The
/// native SDK keeps running but no events leave this provider while consent is
/// denied — documented limitation.
class RollbarProvider: ErrorTrackingProvider {
    let name = "rollbar"
    private(set) var isEnabled: Bool
    private(set) var isInitialized: Bool

    private let config: [String: Any]
    private var debugMode = false
    /// Tracks the consent state. Rollbar has no opt-out toggle, so this provider
    /// gates `logError` itself when consent is revoked.
    private var consentGranted = true

    init(config: [String: Any]) {
        self.config = config
        self.isEnabled = config["enabled"] as? Bool ?? true
        self.isInitialized = false

        guard let accessToken = config["accessToken"] as? String, !accessToken.isEmpty else {
            // No access token -> cannot start Rollbar. Leave uninitialized.
            print("[RollbarProvider] No accessToken provided; Rollbar not started.")
            return
        }

        let environment = config["environment"] as? String ?? "production"

        // NOTE(unverified): build with Xcode — confirm
        // `RollbarConfig.mutableConfig(withAccessToken:environment:)` factory exists
        // (older versions used `RollbarConfiguration` and a different naming scheme)
        // and that `Rollbar.initWithConfiguration(_:)` class method exists in
        // rollbar-ios ~3.2.0. Some 3.x versions use
        // `Rollbar.initWithAccessToken(_:configuration:)` instead.
        let rollbarConfig = RollbarConfig.mutableConfig(withAccessToken: accessToken,
                                                         environment: environment)
        Rollbar.initWithConfiguration(rollbarConfig)

        self.isInitialized = true
    }

    func logError(_ error: Error, context: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted else { return }
        // NOTE(unverified): build with Xcode — confirm
        // `Rollbar.errorError(_:data:context:)` class method exists in rollbar-ios ~3.2.0.
        // Some 3.x versions expose `Rollbar.error(_:data:context:)` or expect the
        // shared notifier via `Rollbar.currentNotifier()` instead.
        Rollbar.errorError(error, data: context, context: "logError")
    }

    func setUserContext(_ userId: String, properties: [String: Any]) {
        guard isEnabled && isInitialized else { return }
        // NOTE(unverified): build with Xcode — confirm
        // `Rollbar.currentConfiguration()` returns a mutable config and that
        // `setPersonId(_:username:email:)` is the current signature in rollbar-ios ~3.2.0.
        // Some 3.x versions expose `currentConfiguration().person = RollbarPerson(...)`.
        let username = properties["username"] as? String
        let email = properties["email"] as? String
        Rollbar.currentConfiguration().setPersonId(userId, username: username, email: email)
    }

    func setConsent(_ granted: Bool) {
        consentGranted = granted
        // NOTE: Rollbar has no built-in opt-out toggle. We gate `logError` at the
        // provider level via `consentGranted`. No SDK call is required here.
    }

    func reset() {
        guard isInitialized else { return }
        // Clear the person attribution on the running Rollbar configuration.
        // NOTE(unverified): build with Xcode — confirm passing `nil` person identity
        // is accepted by `setPersonId(_:username:email:)` in rollbar-ios ~3.2.0.
        // If the SDK rejects nil, an alternative is
        // `Rollbar.currentConfiguration().person = nil`.
        Rollbar.currentConfiguration().setPersonId(nil, username: nil, email: nil)
    }

    func setDebugMode(_ enabled: Bool) {
        // Rollbar's verbose logging is configured at SDK start time via the
        // RollbarConfig object. We retain the flag here for parity; flipping it
        // at runtime is not exposed as a first-class API on the shared notifier.
        debugMode = enabled
    }
}

File: android/src/main/java/com/aoneahsan/unifiedtracking/providers/errortracking/RollbarProvider.java
package com.aoneahsan.unifiedtracking.providers.errortracking;

import android.content.Context;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.aoneahsan.unifiedtracking.providers.ErrorTrackingProvider;
import com.rollbar.android.Rollbar;

import java.util.HashMap;
import java.util.Map;

/**
 * Rollbar error-tracking provider for Android.
 *
 * Uses {@code com.rollbar.android.Rollbar} from the Rollbar Android SDK. The SDK is
 * initialized via {@code Rollbar.init(context, accessToken, environment)} and the
 * singleton is retrieved through {@code Rollbar.instance()}.
 *
 * NOTE(unverified): the build.gradle declares the dependency as
 * {@code com.rollbar:rollbar-java:$rollbarVersion} — but Rollbar publishes a
 * separate {@code com.rollbar:rollbar-android} artifact that contains the
 * {@code com.rollbar.android.Rollbar} class used here. Confirm the correct
 * artifact id in Android Studio; if the project should remain on
 * rollbar-java only, this provider will need to be rewritten to use the
 * generic {@code com.rollbar.notifier.Rollbar} class with manually configured
 * person/server/notifier sections.
 *
 * NOTE(unverified): built against Rollbar Android SDK 1.12.x API. Confirm
 * exact signatures (Rollbar.init overloads, instance().error(Throwable, Map, String),
 * instance().setPersonData(id, username, email), instance().clearPersonData())
 * against the resolved SDK version.
 */
public class RollbarProvider implements ErrorTrackingProvider {
    private static final String TAG = "Rollbar";
    private final Context context;
    private final JSObject config;
    private boolean enabled;
    private boolean initialized;
    private boolean debugMode = false;

    public RollbarProvider(Context context, JSObject config) {
        this.context = context;
        this.config = config;
        this.enabled = config.getBoolean("enabled", true);
        this.initialized = false;

        try {
            String accessToken = config.getString("accessToken");
            if (accessToken == null || accessToken.trim().isEmpty()) {
                Log.e(TAG, "Rollbar accessToken is missing; provider will be inert.");
                return;
            }

            String environment = config.getString("environment", "production");

            // NOTE(unverified): Rollbar.init(Context, String, String) is the documented
            // 3-arg initializer in Rollbar Android 1.x. There are additional overloads
            // that accept registerExceptionHandler/sendUncaught flags; confirm against
            // the resolved SDK.
            Rollbar.init(context, accessToken, environment);
            this.initialized = true;

            if (debugMode) {
                Log.d(TAG, "Rollbar initialized (environment=" + environment + ")");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Rollbar", e);
            this.initialized = false;
        }
    }

    @Override
    public String getName() {
        return "rollbar";
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public boolean isInitialized() {
        return initialized;
    }

    @Override
    public void logError(Exception error, Map<String, Object> errorContext) {
        if (!enabled || !initialized || error == null) return;

        try {
            Rollbar rollbar = Rollbar.instance();
            if (rollbar == null) {
                if (debugMode) {
                    Log.w(TAG, "Rollbar.instance() returned null; skipping notify.");
                }
                return;
            }

            // Defensive copy so we never hand the SDK a null custom-data map.
            Map<String, Object> custom = errorContext != null
                ? new HashMap<>(errorContext)
                : new HashMap<>();

            // NOTE(unverified): Rollbar.error(Throwable, Map<String,Object>, String)
            // is the documented overload (throwable, custom data, description) in
            // Rollbar Android 1.x. The 2-arg error(Throwable, Map) overload also
            // exists on some builds. Confirm in Android Studio.
            rollbar.error(error, custom, null);

            if (debugMode) {
                Log.d(TAG, "Notified Rollbar: " + error.getMessage());
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to notify Rollbar", e);
        }
    }

    @Override
    public void setUserContext(String userId, Map<String, Object> properties) {
        if (!enabled || !initialized) return;

        try {
            Rollbar rollbar = Rollbar.instance();
            if (rollbar == null) return;

            String username = null;
            String email = null;
            if (properties != null) {
                Object u = properties.get("username");
                if (u != null) username = String.valueOf(u);
                Object e = properties.get("email");
                if (e != null) email = String.valueOf(e);
            }

            // NOTE(unverified): Rollbar.setPersonData(String id, String username,
            // String email) is the documented signature in Rollbar Android 1.x.
            // Confirm against the resolved SDK.
            rollbar.setPersonData(userId, username, email);

            if (debugMode) {
                Log.d(TAG, "Set Rollbar person data: " + userId);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set user context", e);
        }
    }

    @Override
    public void setConsent(boolean granted) {
        if (!initialized) return;

        try {
            // Rollbar Android does not expose a runtime opt-out toggle; we gate
            // dispatch locally via the `enabled` flag (logError early-returns when
            // false). The SDK itself stays initialized so consent can be re-granted
            // without re-initializing.
            this.enabled = granted;

            if (debugMode) {
                Log.d(TAG, "Set consent (local gate): " + granted);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set consent", e);
        }
    }

    @Override
    public void reset() {
        if (!initialized) return;

        try {
            Rollbar rollbar = Rollbar.instance();
            if (rollbar == null) return;

            // NOTE(unverified): Rollbar.clearPersonData() is the documented person-data
            // reset in Rollbar Android 1.x; some older SDKs only allow setPersonData(
            // null, null, null). Try clearPersonData first and fall back if it isn't
            // available on the resolved SDK.
            try {
                rollbar.clearPersonData();
            } catch (Throwable t) {
                try {
                    rollbar.setPersonData(null, null, null);
                } catch (Throwable ignored) {
                    // Optional best-effort reset.
                }
            }

            if (debugMode) {
                Log.d(TAG, "Reset Rollbar person data");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to reset", e);
        }
    }

    @Override
    public void setDebugMode(boolean enabled) {
        this.debugMode = enabled;
        if (enabled) {
            Log.d(TAG, "Debug mode enabled for Rollbar provider");
        }
    }
}
